src/Microsoft.AspNet.FileProviders.Combined/CombinedFileProvider.cs

```diff
@@ -0,0 +1,193 @@
+        /// The change token will be notified when one of the change token returned by the provided <see cref="IFileProvider"/> will be notified.</returns>
+        public IChangeToken Watch(string pattern)
+        {
+            // Watch all file providers
+            var activeChangeTokens = new List<IChangeToken>();
+            foreach (var fileProvider in _fileProviders)
+            {
+                var changeToken = fileProvider.Watch(pattern);
+                if (changeToken != null && changeToken.ActiveChangeCallbacks)
+                {
+                    activeChangeTokens.Add(changeToken);
+                }
+            }
+
+            // There is no change token with active change callbacks
+            if (activeChangeTokens.Count == 0)
+            {
+                return NoopChangeToken.Singleton;
+            }
+            var combinedFileChangeToken = new CombinedFileChangeToken(activeChangeTokens);
+            return combinedFileChangeToken;
+        }
+
+        private class CombinedDirectoryContents : IDirectoryContents
+        {
+            private readonly Lazy<Dictionary<string, IFileInfo>> _files;
+
+            public CombinedDirectoryContents(List<IDirectoryContents> listOfFiles)
+            {
+                _files = new Lazy<Dictionary<string, IFileInfo>>(() =>
+                {
+                    var directoryFiles = new Dictionary<string, IFileInfo>();
+                    foreach (var files in listOfFiles)
+                    {
+                        foreach (var file in files)
+                        {
+                            if (!directoryFiles.ContainsKey(file.Name))
+                            {
+                                directoryFiles.Add(file.Name, file);
+                            }
+                        }
+                    }
+                    return directoryFiles;
+                });
+            }
+
+            public IEnumerator<IFileInfo> GetEnumerator()
+            {
+                return _files.Value.Values.GetEnumerator();
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return _files.Value.Values.GetEnumerator();
+            }
+
+            // This directory exists because it is created only when there is existing IDrectoryContents to merge.
+            public bool Exists { get { return true; } }
+        }
+
+        private class CombinedFileChangeToken : IChangeToken
+        {
+            private readonly List<IChangeToken> _changeTokens;
+
+            public CombinedFileChangeToken(List<IChangeToken> changeTokens)
+            {
+                _changeTokens = changeTokens ?? new List<IChangeToken>();
+            }
+
+            public IDisposable RegisterChangeCallback(Action<object> callback, object state)
+            {
+                var disposables = new List<IDisposable>();
+                foreach (var changeToken in _changeTokens)
```

Comment:
Note: when using a list, the compiler doesn't apply any optimization. If you want to avoid the enumerator allocation here, you have to use a good old `for` loop.